bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java

```diff
@@ -49,81 +51,76 @@
 
     /**
      * A private object to implement the MqttCallback interface.
      * We don't want the MqttBrokerConnection to implement this directly.
-     *
+     * Developer hint: Unfortunately MqttCallback does not care about annotations for Null-correctness, we use local
+     * variables to compensate.
      */
     private class ClientCallbacks implements MqttCallback {
         @Override
-        public synchronized void connectionLost(Throwable t) {
-            if (t instanceof MqttException) {
-                MqttException e = (MqttException) t;
-                logger.info("MQTT connection to '{}' was lost: {} : ReasonCode {} : Cause : {}", getName(),
-                        e.getMessage(), e.getReasonCode(),
-                        (e.getCause() == null ? "Unknown" : e.getCause().getMessage()));
-            } else {
-                logger.info("MQTT connection to '{}' was lost: {}", getName(), t.getMessage());
+        public synchronized void connectionLost(@Nullable Throwable exception) {
+            if (exception instanceof MqttException) {
+                MqttException e = (MqttException) exception;
+                logger.info("MQTT connection to '{}' was lost: {} : ReasonCode {} : Cause : {}", host, e.getMessage(),
+                        e.getReasonCode(), (e.getCause() == null ? "Unknown" : e.getCause().getMessage()));
+            } else if (exception != null) {
+                logger.info("MQTT connection to '{}' was lost: {}", host, exception.getMessage());
             }
 
-            for (final MqttConnectionObserver connectionObserver : connectionObservers) {
-                connectionObserver.connectionStateChanged(MqttConnectionState.DISCONNECTED, t);
+            connectionObservers.forEach(o -> o.connectionStateChanged(MqttConnectionState.DISCONNECTED, exception));
+            if (reconnectStrategy != null) {
+                reconnectStrategy.lostConnection();
             }
-
-            reconnectStrategy.lostConnection();
         }
 
         @Override
-        public void deliveryComplete(IMqttDeliveryToken token) {
+        public void deliveryComplete(@Nullable IMqttDeliveryToken token_) {
+            IMqttDeliveryToken token = (@NonNull IMqttDeliveryToken) token_;
             logger.trace("Message with id {} delivered.", token.getMessageId());
         }
 
         @Override
-        public void messageArrived(String topic, MqttMessage message) {
-            logger.trace("Received message on topic '{}' : {}", topic, new String(message.getPayload()));
-            for (Map.Entry<String, List<MqttMessageSubscriber>> entry : consumers.entrySet()) {
-                final String target = entry.getKey();
-                final List<MqttMessageSubscriber> consumerList = entry.getValue();
-
+        public void messageArrived(@Nullable String topic_, @Nullable MqttMessage message_) {
+            byte[] payload = ((@NonNull MqttMessage) message_).getPayload();
```

Comment:
Okay, I get what you are doing here. Instead of defining the methods args `@Nullable` and do the cast here you may add `@NonNullByDefault({})` at `ClientCallbacks` class level. This will "switch off" null annotation checks in this class. It hurts the eyes a little less then casting @Nullabel to @NonNull.